diff --git a/acse/Acse.lex b/acse/Acse.lex
index 663b2db..1687074 100644
--- a/acse/Acse.lex
+++ b/acse/Acse.lex
@@ -92,6 +92,7 @@ ID       [a-zA-Z_][a-zA-Z0-9_]*
 "else"            { return ELSE; }
 "for"             { return FOR; }
 "if"              { return IF; }
+"iif"             { return IIF; }
 "int"             { yylval.intval = INTEGER_TYPE; return TYPE; }
 "while"           { return WHILE; }
 "return"          { return RETURN; }
diff --git a/acse/Acse.y b/acse/Acse.y
index 57ff107..1c80cca 100644
--- a/acse/Acse.y
+++ b/acse/Acse.y
@@ -127,7 +127,7 @@ extern int yyerror(const char* errmsg);
 
 %token <label> DO
 %token <while_stmt> WHILE
-%token <label> IF
+%token <label> IF IIF
 %token <label> ELSE
 %token <intval> TYPE
 %token <svalue> IDENTIFIER
@@ -351,8 +351,20 @@ if_stmt  :  IF
                      gen_beq_instruction (program, $1, 0);
                }
                code_block { $$ = $1; }
+            | IIF 
+              {
+                 $1 = newLabel(program);
+              }
+              LPAR assign_statement SEMI exp RPAR 
+              {
+                 if($6.expression_type == IMMEDIATE)
+                    gen_load_immediate(program, $6.value);
+                 else
+                    gen_andb_instruction(program, $6.value, $6.value, $6.value, CG_DIRECT_ALL);
+                 gen_beq_instruction(program, $1, 0);
+              }
+              code_block {$$ = $1;}
 ;
-
 while_statement  : WHILE
                   {
                      /* initialize the value of the non-terminal */
diff --git a/tests/Makefile b/tests/Makefile
index ebbebd6..65c6186 100644
--- a/tests/Makefile
+++ b/tests/Makefile
@@ -1,7 +1,7 @@
 asm=$(CURDIR)/../bin/asm
 acse=$(CURDIR)/../bin/acse
 export asm acse
-dirs:=arrays sum dowhile mult_table power semi assign arrays arrays2 misc
+dirs:=arrays sum dowhile mult_table power semi assign arrays arrays2 misc iif
 
 test :
 	for i in $(dirs); do $(MAKE) -C $$i -f ../Makefile.test ; done
diff --git a/tests/iif/iif.src b/tests/iif/iif.src
new file mode 100644
index 0000000..ae99b23
--- /dev/null
+++ b/tests/iif/iif.src
@@ -0,0 +1,11 @@
+int i,j;
+
+i=5;
+
+j=7;
+
+iif (i=i+2; (j*5)>i ){
+  j = i;
+}
+
+write(j);
